### PR TITLE
Treasury Vault v2

### PR DESCRIPTION
**Pull request by banteg**
Added current TreasuryVault v1 contract from [`0x93A62dA5a14C80f265DAbC077fCEE437B1a0Efde`](https://etherscan.io/address/0x93A62dA5a14C80f265DAbC077fCEE437B1a0Efde#code)

Reworking it to implement changes [specified here](https://gov.yearn.finance/t/yip-upgrade-treasury-vault-to-solve-gas-subsidy-gitcoin-and-more/4732).

### CODE REVIEW

**Comment by banteg** [comment]
Curve will deploy new registry soon, ideally we tap into it and not maintain the registry ourselves.

File: contracts/utility/TreasuryVault.sol [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        uint[] calldata coins
+    ) external {
+        require(msg.sender == governance, "!governance");
+        for (uint i = 0; i < tokens.length; i++) {
+            curvePools[tokens[i]] = CurvePool(pools[i], coins[i]);
+        }
+    }
+
+    function withdrawCurve(address[] calldata tokens) external {
+        require(authorized[msg.sender], "!authorized");
```
Comment (dmolina79) [inline contracts/utility/TreasuryVault.sol:105]: `authorized[msg.sender] == true` like line 120?

File: contracts/utility/TreasuryVault.sol [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            if (balance == 0) continue;
+            if (pool.coins == 2) ICurveExchange(pool.swap).remove_liquidity(balance, [uint256(0), 0]);
+            if (pool.coins == 3) ICurveExchange(pool.swap).remove_liquidity(balance, [uint256(0), 0, 0]);
+            if (pool.coins == 4) ICurveExchange(pool.swap).remove_liquidity(balance, [uint256(0), 0, 0, 0]);
+        }
+    }
+
+    // Only allows to withdraw non-core strategy tokens ~ this is over and above normal yield
+    function convert(address _from, uint parts) external {
+        require(authorized[msg.sender] == true, "!authorized");
```
Comment (gustavoguimaraes) [inline contracts/utility/TreasuryVault.sol:120]: unnecessary boolean check.

**Comment by banteg** [comment]
I think this can be obsoleted. In v2 the Treasury Vault can be structured like a Vault with multiple zero-debt strategies for each reward token, every one of which just converts the rewards to `want`.

**Comment by fubuloubu** [comment]
Well, if this is deployed on chain then it should be kept up to date

**Pull request closed without merging**